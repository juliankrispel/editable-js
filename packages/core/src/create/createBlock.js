// @flow

import type { RawBlock, Block } from '../types'
import createId from './createId'
import createCharacterData from './createCharacterData'

export default function createBlock (block: RawBlock): Block {
  const { children, text = '', key, characterRanges, ...rest } = block

  if (process.env.NODE_ENV !== 'production' && key != null && !isNaN(Number(key))) {
    throw new Error('block key can not be a number')
  }

  const newBlock = {
    ...rest,
    text,
    characterData: createCharacterData(block),
    children: [],
    key: key || createId()
  }

  if (Array.isArray(children)) {
    newBlock.children = children.map(createBlock)
  }

  return newBlock
}
