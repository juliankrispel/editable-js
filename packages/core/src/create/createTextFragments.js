// @flow
import type { TextFragment, Block, EntityMap } from '../types'
import hasEqualCharacterData from '../hasEqualCharacterData'

export default function createTextFragments(block: Block, entityMap: EntityMap): Array<TextFragment> {
  return block.characterData.reduce(
    (acc, data, index) => {
      if (acc.length < 1) {
        return [{
          marks: data.marks,
          entity: data.entity,
          offset: index,
          text: block.text[index]
        }]
      } else {
        const lastFragment = acc[acc.length - 1]
        if (hasEqualCharacterData(lastFragment, data)) {
          return acc.slice(0, -1).concat([{
            ...lastFragment,
            text: lastFragment.text + block.text[index]
          }])
        } else {
          return [
            ...acc,
            {
              marks: data.marks,
              entity: data.entity,
              offset: index,
              text: block.text[index]
            }
          ]
        }
      }
    },
    []
  ).map(fragment => ({
    ...fragment,
    entity: fragment.entity != null ? entityMap[fragment.entity] : null
  }))
}
