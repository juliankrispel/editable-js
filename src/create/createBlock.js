// @flow

import type { RawBlock, Block } from '../types'
import createId from './createId'

export default function createBlock (block: RawBlock): Block {
  const { children, key, ...rest } = block

  const newBlock = {
    ...rest,
    //    characterData: Array(block.text.length).fill([]),
    children: [],
    key: key || createId()
  }

  if (Array.isArray(children)) {
    newBlock.children = children.map(createBlock)
  }

  return newBlock
}
