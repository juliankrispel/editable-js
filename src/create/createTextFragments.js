// @flow
import type { TextFragment, Block, EntityMap } from '../types'

const hasEqualMarks = (left, right) =>
  right.entity === left.entity &&
  Array.from(left.marks).sort().join('') === Array.from(right.marks).sort().join('')

export default function createTextFragments(block: Block, entityMap: EntityMap): Array<TextFragment> {
  return block.characterData.reduce(
    (acc, data, index) => {
      if (acc.length < 1) {
        return [{
          marks: data.marks,
          entity: data.entity,
          offset: index,
          text: block.text[index]
        }]
      } else {
        const lastFragment = acc[acc.length - 1]
        if (hasEqualMarks(lastFragment, data)) {
          return acc.slice(0, -1).concat([{
            ...lastFragment,
            text: lastFragment.text + block.text[index]
          }])
        } else {
          return [
            ...acc,
            {
              marks: data.marks,
              entity: data.entity,
              offset: index,
              text: block.text[index]
            }
          ]
        }
      }
    },
    []
  ).map(fragment => ({
    ...fragment,
    entity: fragment.entity != null ? entityMap[fragment.entity] : null,
  }))
}
