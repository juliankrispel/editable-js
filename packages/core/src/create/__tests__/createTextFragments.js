// @flow

import createTextFragments from '../createTextFragments'
import createBlock from '../createBlock'

describe('createTextFragments', () => {
  describe('creates text fragments from block element', () => {
    test('without marks', () => {
      const block = createBlock({
        text: 'Hello World'
      })

      const result = createTextFragments(block)

      expect(result).toMatchSnapshot()
    })

    test('with marks', () => {
      const block = createBlock({
        text: 'Hello World',
        characterRanges: [{
          offset: 0,
          length: 4,
          marks: ['bold', 'italic']
        }]
      })

      const result = createTextFragments(block)

      expect(result).toMatchSnapshot()
    })

    test('with overlapping marks', () => {
      const block = createBlock({
        text: 'Hello World',
        characterRanges: [{
          offset: 0,
          length: 4,
          marks: ['bold']
        }, {
          offset: 1,
          length: 6,
          marks: ['italic']
        }, {
          offset: 3,
          length: 9,
          entities: ['1']
        }]
      })

      const result = createTextFragments(block)

      expect(result).toMatchSnapshot()
    })
  })
})
