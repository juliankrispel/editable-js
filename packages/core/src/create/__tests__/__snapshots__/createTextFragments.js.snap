// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTextFragments creates text fragments from block element with marks 1`] = `
Array [
  Object {
    "entity": null,
    "marks": Array [
      "bold",
      "italic",
    ],
    "offset": 0,
    "text": "Hell",
  },
  Object {
    "entity": null,
    "marks": Array [],
    "offset": 4,
    "text": "o World",
  },
]
`;

exports[`createTextFragments creates text fragments from block element with overlapping marks 1`] = `
Array [
  Object {
    "entity": null,
    "marks": Array [
      "bold",
    ],
    "offset": 0,
    "text": "H",
  },
  Object {
    "entity": null,
    "marks": Array [
      "bold",
      "italic",
    ],
    "offset": 1,
    "text": "ell",
  },
  Object {
    "entity": null,
    "marks": Array [
      "italic",
    ],
    "offset": 4,
    "text": "o W",
  },
  Object {
    "entity": null,
    "marks": Array [],
    "offset": 7,
    "text": "orld",
  },
]
`;

exports[`createTextFragments creates text fragments from block element without marks 1`] = `
Array [
  Object {
    "entity": null,
    "marks": Array [],
    "offset": 0,
    "text": "Hello World",
  },
]
`;
